bittide-instances/tests/Wishbone/Watchdog.hs

```diff
@@ -0,0 +1,95 @@
+-- SPDX-FileCopyrightText: 2022 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# OPTIONS_GHC -fplugin=Protocols.Plugin #-}
+
+module Wishbone.Watchdog where
+
+-- Preludes
+import Clash.Prelude
+
+-- Local
+import Bittide.DoubleBufferedRam
+import Bittide.Instances.Domains
+import Bittide.ProcessingElement
+import Bittide.ProcessingElement.Util
+import Bittide.SharedTypes
+import Bittide.Wishbone
+import Project.FilePath
+
+-- Other
+import Data.Char
+import Data.Maybe
+import Language.Haskell.TH
+import Protocols
+import Protocols.Idle
+import System.FilePath
+import Test.Tasty
+import Test.Tasty.HUnit
+import Test.Tasty.TH
+import VexRiscv (DumpVcd (NoDumpVcd))
+
+-- Qualified
+import qualified Data.List as L
+import qualified Protocols.Df as Df
+
+sim :: IO ()
+sim = putStrLn simResult
+
+simResult :: String
+simResult = chr . fromIntegral <$> mapMaybe Df.dataToMaybe uartStream
+ where
+  uartStream = sampleC def dut
+
+{- | Run the timing module self test with processingElement and inspect it's uart output.
+The test returns names of tests and a boolean indicating if the test passed.
+-}
+case_time_rust_self_test :: Assertion
+case_time_rust_self_test = do
+  let result = L.head $ lines simResult
+  assertEqual "Measured timeout wrong " result "Timeout took 50 microseconds"
+
+{- | A simple instance containing just VexRisc and UART as peripheral.
+Runs the `hello` binary from `firmware-binaries`.
+-}
+dut ::
+  Circuit () (Df Basic200 (BitVector 8))
+dut = withClockResetEnable clockGen resetGen enableGen
+  $ circuit
+  $ \_unit -> do
+    (uartRx, jtag) <- idleSource -< ()
+    [uartBus, timeBus, idleBusA, idleBusB] <- processingElement NoDumpVcd peConfig -< jtag
+
+    idleSink <| (watchDogWb @_ @_ @4 "1 cycle" d1) -< idleBusA
+    idleSink
+      <| (watchDogWb @_ @_ @4 "50 us" (SNat @(PeriodToCycles Basic200 (Microseconds 50))))
+      -< idleBusB
+
+    timeWb <| (watchDogWb @_ @_ @4 "" d0) -< timeBus
+    (uartTx, _uartStatus) <- (uartInterfaceWb @_ @_ @4) d2 d2 uartSim -< (uartBus, uartRx)
+    idC -< uartTx
+ where
+  (iMem, dMem) =
+    $( do
+        root <- runIO $ findParentContaining "cabal.project"
+        let
+          elfDir = root </> firmwareBinariesDir "riscv32imc-unknown-none-elf" Release
+          elfPath = elfDir </> "watchdog_test"
+
+          iSize = 64 * 1024 -- 64 KB
+          dSize = 64 * 1024 -- 64 KB
+        memBlobsFromElf BigEndian (Just iSize, Just dSize) elfPath Nothing
+     )
```

Comment:
What do we want?
Fewer dependencies on Rust code in Haskell!
When do we want it?
Now!

https://github.com/bittide/bittide-hardware/blob/325125fc272e6653a89363015fc1c6678715275b/bittide-instances/src/Bittide/Instances/Pnr/Ethernet.hs#L159-L171

Reply:
Its been done